### PR TITLE
Modify `eosio.msig` to allow `eosio.code` permission level authorization

### PR DESCRIPTION
**Pull request by johndebord**
The following precursor checks have been added to actions `approve` and `unapprove` in the `eosio.msig` contract, which allows a contract to utilize `eosio.code` permission level authorization.:

```c++
void multisig::approve( name proposer, name proposal_name, permission_level level,
                        const eosio::binary_extension<eosio::checksum256>& proposal_hash )
{
   if ( level.permission == "eosio.code"_n ) {
       check( get_sender() == level.actor, "wrong contract sent `approve` action for eosio.code permmission" );
   } else {
      require_auth( level );
   }
   ...
   ...
   ...
}
```

```c++
void multisig::unapprove( name proposer, name proposal_name, permission_level level ) {
   if ( level.permission == "eosio.code"_n ) {
       check( get_sender() == level.actor, "wrong contract sent `unapprove` action for eosio.code permmission" );
   } else {
      require_auth( level );
   }
   ...
   ...
   ...
}
```

### CODE REVIEW

File: contracts/test_contracts/sendinline/src/wrongcon.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+   [[eosio::action]]
+   void send( eosio::name contract, eosio::name action_name, std::vector<eosio::permission_level> auths, std::vector<char> payload) {
+      eosio::action act;
+      act.account = contract;
+      act.name = action_name;
+      act.authorization = auths;
+      act.data = std::move(payload);
+      act.send();
+   }
+};
```
Comment (swatanabe-b1) [inline contracts/test_contracts/sendinline/src/wrongcon.cpp]: This looks identical to sendinline.  Is there any reason to have both?

File: tests/contracts.hpp.in [hunk 2]
```diff
@@ -14,16 +14,20 @@ struct contracts {
    static std::vector<char>    wrap_abi() { return read_abi("${CMAKE_BINARY_DIR}/../contracts/eosio.wrap/eosio.wrap.abi"); }
    static std::vector<uint8_t> bios_wasm() { return read_wasm("${CMAKE_BINARY_DIR}/../contracts/eosio.bios/eosio.bios.wasm"); }
    static std::vector<char>    bios_abi() { return read_abi("${CMAKE_BINARY_DIR}/../contracts/eosio.bios/eosio.bios.abi"); }
+   static std::vector<uint8_t> sendinline_wasm() {return read_wasm("${CMAKE_BINARY_DIR}/../contracts/test_contracts/sendinline/sendinline.wasm"); }
+   static std::vector<char>    sendinline_abi() {return read_abi("${CMAKE_BINARY_DIR}/../contracts/test_contracts/sendinline/sendinline.abi"); }
+   static std::vector<uint8_t> wrongcon_wasm() {return read_wasm("${CMAKE_BINARY_DIR}/../contracts/test_contracts/sendinline/wrongcon.wasm"); }
```
Comment (arhag) [inline tests/contracts.hpp.in]: There is no need for the `wrongcon` contract. It is identical to `sendinline` so just use that.

File: tests/contracts.hpp.in [hunk 3]
```diff
@@ -14,16 +14,20 @@ struct contracts {
    static std::vector<char>    wrap_abi() { return read_abi("${CMAKE_BINARY_DIR}/../contracts/eosio.wrap/eosio.wrap.abi"); }
    static std::vector<uint8_t> bios_wasm() { return read_wasm("${CMAKE_BINARY_DIR}/../contracts/eosio.bios/eosio.bios.wasm"); }
    static std::vector<char>    bios_abi() { return read_abi("${CMAKE_BINARY_DIR}/../contracts/eosio.bios/eosio.bios.abi"); }
+   static std::vector<uint8_t> sendinline_wasm() {return read_wasm("${CMAKE_BINARY_DIR}/../contracts/test_contracts/sendinline/sendinline.wasm"); }
```
Comment (arhag) [inline tests/contracts.hpp.in]: I don't know why we called the struct `contracts::util` (it probably should have been `contracts::test`), but it seems we put all contracts that are part of the actual eosio.contracts product in `contracts` and all the remaining contracts needed for testing into `contracts::util`. So to continue that convention, please move these two functions for the `sendinline` contract to `util`.

File: tests/eosio.msig_tests.cpp [hunk 4]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+   create_accounts( {N(sendinline)} );
+   set_code( N(sendinline), contracts::contracts::sendinline_wasm() );
+   set_abi( N(sendinline), contracts::contracts::sendinline_abi().data() );
+
+   create_accounts( {N(wrongcon)} );
+   set_code( N(wrongcon), contracts::contracts::wrongcon_wasm() );
+   set_abi( N(wrongcon), contracts::contracts::wrongcon_abi().data() );
+   produce_blocks();
+
+   static action act = get_action(config::system_account_name, N(reqauth), {}, mvo()("from", N(alice)));
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: Remove `static`.

File: tests/eosio.msig_tests.cpp [hunk 5]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+   BOOST_REQUIRE_EQUAL( N(alice), name{trx_trace->action_traces.at(1).act.authorization[0].actor} );
+   BOOST_REQUIRE_EQUAL( N(perm), name{trx_trace->action_traces.at(1).act.authorization[0].permission} );
+
+   act = get_action(N(eosio.msig), N(approve), {}, mvo()
+                     ("proposer", N(bob))
+                     ("proposal_name", N(first))
+                     ("level", permission_level{N(sendinline), N(eosio.code)})
+   );
+   
+   static transaction trx = reqauth( N(alice), {permission_level{N(alice), N(perm)}}, abi_serializer_max_time );
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: Remove `static`.

File: tests/eosio.msig_tests.cpp [hunk 6]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+   
+   base_tester::push_action( N(eosio.msig), N(propose), N(bob), mvo()
+                              ("proposer", "bob")
+                              ("proposal_name", "first")
+                              ("trx", trx)
+                              ("requested", std::vector<permission_level>{{ N(sendinline), N(eosio.code) }})
+   );
+   BOOST_REQUIRE_EQUAL( false, get_earliest_exec_time(N(bob), N(first)).valid() );
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: You can remove this check.

File: tests/eosio.msig_tests.cpp [hunk 7]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+                              ("contract", N(eosio.msig))
+                              ("action_name", "approve")
+                              ("auths", std::vector<permission_level>{})
+                              ("payload", act.data)
+                            ),
+                            eosio_assert_message_exception,
+                            eosio_assert_message_is("wrong contract sent `approve` action for eosio.code permmission")
+   );
+   
+   // `approve` shall succeed when being sent from the wrong contract
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: You meant to say "... from the right contract".

File: tests/eosio.msig_tests.cpp [hunk 8]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+                              ("contract", N(eosio.msig))
+                              ("action_name", "unapprove")
+                              ("auths", std::vector<permission_level>{})
+                              ("payload", act.data)
+                            ),
+                            eosio_assert_message_exception,
+                            eosio_assert_message_is("wrong contract sent `unapprove` action for eosio.code permmission")
+   );
+   
+   // `unapprove` shall fail when being sent from the wrong contract
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: You meant to say "... from the right contract".

File: tests/eosio.msig_tests.cpp [hunk 9]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+   BOOST_REQUIRE_EQUAL( fc::unsigned_int{2}, trx_trace->action_traces.at(1).action_ordinal );
+   BOOST_REQUIRE_EQUAL( fc::unsigned_int{1}, trx_trace->action_traces.at(1).creator_action_ordinal );
+   BOOST_REQUIRE_EQUAL( fc::unsigned_int{1}, trx_trace->action_traces.at(1).closest_unnotified_ancestor_action_ordinal );
+   BOOST_REQUIRE_EQUAL( N(eosio), action_name{trx_trace->action_traces.at(1).receiver} );
+   BOOST_REQUIRE_EQUAL( N(eosio), name{trx_trace->action_traces.at(1).act.account} );
+   BOOST_REQUIRE_EQUAL( N(reqauth), name{trx_trace->action_traces.at(1).act.name} );
+   BOOST_REQUIRE_EQUAL( N(alice), name{trx_trace->action_traces.at(1).act.authorization[0].actor} );
+   BOOST_REQUIRE_EQUAL( N(perm), name{trx_trace->action_traces.at(1).act.authorization[0].permission} );
+
+   act = get_action(N(eosio.msig), N(approve), {}, mvo()
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: Even though the payload is the same for both `approve` and `unapprove` actions in the tests below, you should still split this into a `approve_act` and `unapprove_act` (using the appropriate action to generate the action payload for each) and then use the appropriate `data` field in the checks below.

File: tests/eosio.msig_tests.cpp [hunk 10]
```diff
@@ -1366,4 +1366,170 @@ BOOST_FIXTURE_TEST_CASE( unnecessary_approve_logic, eosio_msig_tester ) try {
...
+   // `approve` to get back into a state ready for `exec`
+   base_tester::push_action( N(sendinline), N(send), N(bob), mvo()
+                              ("contract", N(eosio.msig))
+                              ("action_name", "approve")
+                              ("auths", std::vector<permission_level>{})
+                              ("payload", act.data)
+   );
+   produce_blocks();
+   
+   trx_trace = base_tester::push_action( N(eosio.msig), N(exec), N(alice), mvo()
```
Comment (arhag) [inline tests/eosio.msig_tests.cpp]: Use `bob` as the `executer` as well.

**Comment by arhag** [comment]
Also please update the title and add a description to reflect the fact that is doing more than just moving contracts around and adding a new test. It is also augmenting the behavior of the `eosio.msig` contract to enable contract to call `approve` and `unapprove` with a permission level that is the `eosio.code` permission of the sending contract.

File: contracts/test_contracts/sendinline/src/wrongcon.cpp (hunk 1 above)
Comment (johndebord) [reply contracts/test_contracts/sendinline/src/wrongcon.cpp]: Removed file `wrongcon.cpp`.

File: tests/contracts.hpp.in (hunk 2 above)
Comment (johndebord) [reply tests/contracts.hpp.in]: Removed file `wrongcon.cpp`.

File: tests/contracts.hpp.in (hunk 3 above)
Comment (johndebord) [reply tests/contracts.hpp.in]: Moved to `util` namespace.

File: tests/eosio.msig_tests.cpp (hunk 4 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Removed `static`.

File: tests/eosio.msig_tests.cpp (hunk 5 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Removed `static`.

File: tests/eosio.msig_tests.cpp (hunk 6 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Removed check.

File: tests/eosio.msig_tests.cpp (hunk 7 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Corrected comment.

File: tests/eosio.msig_tests.cpp (hunk 8 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Corrected comment.

File: tests/eosio.msig_tests.cpp (hunk 9 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Separated `act` into `approve_act` and `unapprove_act`.

File: tests/eosio.msig_tests.cpp (hunk 10 above)
Comment (johndebord) [reply tests/eosio.msig_tests.cpp]: Now using `bob` as the `executer`.

**Pull request merged**